File: backend/app/schemas/provider.py
import uuid
from datetime import datetime, date, time
from decimal import Decimal
from typing import Optional, List
from pydantic import BaseModel, Field, ConfigDict


# ==========================================
# PROVIDER SCHEMAS
# ==========================================

class ProviderBase(BaseModel):
    full_name: str = Field(..., max_length=255)
    photo_url: Optional[str] = Field(None, max_length=1024)
    category: Optional[str] = Field(None, max_length=100)
    skills: Optional[str] = None
    experience_years: int = Field(0, ge=0)
    base_price: Decimal = Field(Decimal("0.00"), ge=Decimal("0.00"))
    service_area: Optional[str] = Field(None, max_length=255)


class ProviderProfileUpdate(BaseModel):
    full_name: Optional[str] = Field(None, max_length=255)
    photo_url: Optional[str] = Field(None, max_length=1024)
    category: Optional[str] = Field(None, max_length=100)
    skills: Optional[str] = None
    experience_years: Optional[int] = Field(None, ge=0)
    base_price: Optional[Decimal] = Field(None, ge=Decimal("0.00"))
    service_area: Optional[str] = Field(None, max_length=255)


class ProviderProfileResponse(ProviderBase):
    user_id: uuid.UUID
    is_verified: bool
    reliability_score: Decimal
    acceptance_rate: Decimal
    cancellation_rate: Decimal
    no_show_rate: Decimal
    on_time_rate: Decimal
    response_time_score: Decimal
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


class ProviderMetricsResponse(BaseModel):
    user_id: uuid.UUID
    reliability_score: Decimal
    acceptance_rate: Decimal
    cancellation_rate: Decimal
    no_show_rate: Decimal
    on_time_rate: Decimal
    response_time_score: Decimal

    model_config = ConfigDict(from_attributes=True)


# ==========================================
# CERTIFICATE SCHEMAS
# ==========================================

class CertificateCreate(BaseModel):
    document_url: str = Field(..., max_length=1024, description="URL to the uploaded certificate document")
    certificate_type: str = Field(..., max_length=100, description="Type of document, e.g., Licence, ID, Diploma")


class CertificateResponse(BaseModel):
    id: uuid.UUID
    provider_id: uuid.UUID
    document_url: str
    certificate_type: str
    verification_status: str
    verified_by: Optional[uuid.UUID] = None
    uploaded_at: datetime
    verified_at: Optional[datetime] = None

    model_config = ConfigDict(from_attributes=True)


# ==========================================
# AVAILABILITY SCHEMAS
# ==========================================

class AvailabilityCreate(BaseModel):
    slot_date: date = Field(..., description="Date of the available slot")
    start_time: time = Field(..., description="Start time of the slot")
    end_time: time = Field(..., description="End time of the slot")


class AvailabilityResponse(BaseModel):
    id: uuid.UUID
    provider_id: uuid.UUID
    slot_date: date
    start_time: time
    end_time: time
    status: str
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)


# ==========================================
# PROVIDER SERVICE SCHEMAS
# ==========================================

class ProviderServiceCreate(BaseModel):
    service_id: uuid.UUID = Field(..., description="Target service UUID")
    price: Decimal = Field(..., ge=Decimal("0.00"), description="Service customized rate")
    duration_minutes: int = Field(60, ge=15, description="Estimated duration in minutes")
    active: bool = Field(True, description="Active status of this service offering")


class ProviderServiceUpdate(BaseModel):
    price: Optional[Decimal] = Field(None, ge=Decimal("0.00"))
    duration_minutes: Optional[int] = Field(None, ge=15)
    active: Optional[bool] = None


class ProviderServiceResponse(BaseModel):
    id: uuid.UUID
    provider_id: uuid.UUID
    service_id: uuid.UUID
    price: Decimal
    duration_minutes: int
    active: bool
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)

File: frontend/src/types/provider.ts
export interface ProviderProfile {
  user_id: string;
  full_name: string;
  photo_url?: string | null;
  category?: string | null;
  skills?: string | null;
  experience_years: number;
  base_price: string | number;
  service_area?: string | null;
  is_verified: boolean;
  reliability_score: number;
  acceptance_rate: number;
  cancellation_rate: number;
  no_show_rate: number;
  on_time_rate: number;
  response_time_score: number;
  created_at: string;
  updated_at: string;
}

export interface ProviderProfileUpdate {
  full_name?: string;
  photo_url?: string;
  category?: string;
  skills?: string;
  experience_years?: number;
  base_price?: string | number;
  service_area?: string;
}

export interface Certificate {
  id: string;
  provider_id: string;
  document_url: string;
  certificate_type: string;
  verification_status: 'PENDING' | 'VERIFIED' | 'REJECTED';
  verified_by?: string | null;
  uploaded_at: string;
  verified_at?: string | null;
}

export interface CertificateCreate {
  document_url: string;
  certificate_type: string;
}

export interface AvailabilitySlot {
  id: string;
  provider_id: string;
  slot_date: string;
  start_time: string;
  end_time: string;
  status: string;
  created_at: string;
}

export interface AvailabilityCreate {
  slot_date: string;
  start_time: string;
  end_time: string;
}

export interface ProviderServiceOffering {
  id: string;
  provider_id: string;
  service_id: string;
  price: string | number;
  duration_minutes: number;
  active: boolean;
  created_at: string;
}

export interface ProviderServiceCreate {
  service_id: string;
  price: string | number;
  duration_minutes?: number;
  active?: boolean;
}

export interface ProviderServiceUpdate {
  price?: string | number;
  duration_minutes?: number;
  active?: boolean;
}
